Tests/MongoSwiftTests/ServerSelectionTests.swift

```diff
@@ -6,62 +6,135 @@ import TestsCommon
         ])
-        let replicaSetNoPrimarySuitableServers = replicaSetNoPrimaryTopology
-            .findSuitableServers(readPreference: primaryReadPreference)
-        expect(replicaSetNoPrimarySuitableServers).to(haveCount(0))
+        let suitable1 = try replicaSetNoPrimaryTopology
+            .findSuitableServers(readPreference: self.primaryReadPreference)
+        expect(suitable1).to(haveCount(0))
 
-        let replicaSetNoPrimarySuitableServer2 = replicaSetNoPrimaryTopology.findSuitableServers(readPreference: nil)
-        expect(replicaSetNoPrimarySuitableServer2).to(haveCount(0))
+        let suitable2 = try replicaSetNoPrimaryTopology
+            .findSuitableServers(readPreference: nil)
+        expect(suitable2).to(haveCount(0))
 
-        let replicaSetNoPrimarySuitableServer3 = replicaSetNoPrimaryTopology
-            .findSuitableServers(readPreference: primaryPrefReadPreferemce)
-        expect(replicaSetNoPrimarySuitableServer3[0].type).to(equal(.rsSecondary))
-        expect(replicaSetNoPrimarySuitableServer3).to(haveCount(2))
+        let suitable3 = try replicaSetNoPrimaryTopology
+            .findSuitableServers(readPreference: self.primaryPrefReadPreferemce)
+        expect(suitable3[0].type).to(equal(.rsSecondary))
+        expect(suitable3).to(haveCount(2))
+    }
 
-        // sharded
+    func testShardedTopology() throws {
         let shardedTopology = TopologyDescription(type: .sharded, servers: [
             mongosServer
         ])
-        let shardedSuitableServers = shardedTopology.findSuitableServers()
+        let shardedSuitableServers = try shardedTopology.findSuitableServers()
         expect(shardedSuitableServers[0].type)
             .to(equal(.mongos))
         expect(shardedSuitableServers).to(haveCount(1))
     }
+
+    func testTagSets() throws {
+        // tag set 1
+        let topology = TopologyDescription(type: .replicaSetNoPrimary, servers: [
+            rsSecondaryServer1,
+            rsSecondaryServer2,
+            rsSecondaryServer3
+        ])
+        let secondaryReadPreferenceWithTagSet = try ReadPreference(
+            .secondaryPreferred,
+            tagSets: [tagSet],
+            maxStalenessSeconds: nil
+        )
+
+        let suitable = try topology
+            .findSuitableServers(readPreference: secondaryReadPreferenceWithTagSet)
+        expect(suitable[0].type).to(equal(.rsSecondary))
+        expect(suitable).to(haveCount(1))
+
+        // tag set 2
+        let secondaryReadPreferenceWithTagSet2 = try ReadPreference(
+            .secondaryPreferred,
+            tagSets: [tagSet2],
+            maxStalenessSeconds: nil
+        )
+
+        let suitable2 = try topology
+            .findSuitableServers(readPreference: secondaryReadPreferenceWithTagSet2)
+        expect(suitable2[0].type).to(equal(.rsSecondary))
+        expect(suitable2).to(haveCount(2))
+
+        // invalid tag set passing
+        let primaryReadPreferenceWithTagSet = try ReadPreference(
```

Comment:
I was surprised this works but I realized this initializer does not actually validate that tagSets aren't specified when the mode is primary.

could you please add validation for that in the initializer? 

it would prevent us from testing this case but I think that's ok; hitting the new `InternalError` you added is unexpected and points to a bug in our own code somewhere

Reply:
Yeah it makes more sense in the initializer, I added it. I just modified the test to make sure the error is thrown from the initializer